### DISCUSSION TITLE
What is the best way to reuse CSS baseline grid calculations?

### DISCUSSION THREAD

**Discussion by steida** [facebook/stylex#388] [Q&A]
From this awesome article https://maketypework.com/web-typography-baseline-grids-made-easy

**Comment by nmn** [comment]
The answer is to make typography components. Make `<H level={}>`, `<P>`, `<Ol>`, `<Ul>`, `<Li>`, components. Put the grid calculations within them.

-----


Separately, this article looks interesting and I'm spend some more time reading it, but I get the feeling that it's trying to achieve the same thing as capsize.js by by using the `cap` unit. If I'm right, this approach will break down with different fonts. The approach is highly dependent on the ratio of ascender height to descender height in a particular font.

This is exactly what capsize.js helps with.

**Comment by steida** [comment]
@nmn Thank you for the answer. Let me explain the issue I have with the always-component approach.

I need to reuse a style because basic text is rendered by different components. Hence, I have $styles.

```ts
/**
 * Sharing styles is not generally a good idea; components are better, but the
 * text is rendered with different components, a button, input, etc., so
 * sometimes it makes sense.
 */
export const $styles = stylex.create({
  text: {
    color: colors.primary,
    fontFamily: fonts.sans,
    fontSize: fontSizes.step0,
    // https://maketypework.com/web-typography-baseline-grids-made-easy/
    lineHeight: baseline.lineHeight,
    marginBottom: baseline.push,
    paddingTop: baseline.shift,
  },
});
```

In my tokens.stylex.ts

```ts
// https://github.com/facebook/stylex/issues/250#issuecomment-1868392679
// vars, because they are supposed to be set
export const vars = stylex.defineVars({
  buttonPressableBackgroundColor: "transparent",
  baselines: "1",
  baselinesBeneath: "0", // better than marginBottom spacing, because it's within a rhythm
});

// https://maketypework.com/web-typography-baseline-grids-made-easy/
const baselineHeight = fontSizes.step0;
const baselineShift = `calc(calc(calc(${vars.baselines} * ${baselineHeight}) - 1cap) / 2)`;
export const baseline = stylex.defineVars({
  height: baselineHeight,
  lineHeight: `calc(${vars.baselines} * ${baselineHeight})`,
  shift: baselineShift,
  push: `calc(calc(${vars.baselinesBeneath} * ${baselineHeight}) - ${baselineShift})`,
});
```

I read discussions about reusing vars within a file. Am I doing that properly?

**Reply by steida** [reply]
As for "this approach will break down with different fonts."

Are you sure? Isn't a cap unit exactly for that?

There are more authors who recommends it https://blog.kizu.dev/cap-height-align/

**Reply by nmn** [reply]
> Are you sure? Isn't a cap unit exactly for that?

Sadly, yes I'm sure. I tried this more than a year ago at Meta and ran into such problems. `1cap - 1ex` Will always give you the height of the ascender of the font (`a` vs `A`). But there is currently no way to get the height of the descender (`a` vs `g`). 

All of these approaches just assume that the ascender and descender heights are the same, and in many fonts, like Arial, Helvetica, SF UI, Roboto etc. that's close to the truth. (The descender is usually more than 80% as tall as the ascender), this isn't true in other fonts like Bodoni or other fonts with tall Ascenders.

Even in the first example you shared, you can see that the larger text is just a few pixels below where it should be.

The solution to this problem is relatively easy. You just need to know the ascender to descender ratio in the fonts you are using and and you can fix the math. This is exactly what capsize.js does for you. (It's just a tool to generate stuff. Not a runtime library).

-------

Regarging your code, you can still make components and accept as `as` prop to render different components and `style` to mix in custom styles.

You can also just define and export `stylex.create` from a file for `step1`, `step2` etc. These styles can only contain:

```tsx
fontSize: 16,
lineHeight: 1.5,
marginBottom: ...
paddingTop: ...
```

Then you can import and use it wherever you need.

----

Regarding your variable approach, I don't have enough context on how you're using that to say for sure if that would work.

**Reply by predaytor** [reply]
@nmn, is it possible to integrate the @capsizecss stack using StyleX directly into the code with no runtime overhead? Always wanted to try this (if it's even worth it), but it's only recently that design tools like Figma and Safari Working Group (CSS spec for `text-box-trim`) have started to think about this problem, so we may use it in the future. Also, it is difficult (or impossible) to integrate for fluid font sizes.

**Reply by steida** [reply]
@nmn Are you sure you are using the right terminology? Arial ascender and descender are not almost the same for Arial.

![Screenshot 2024-01-28 at 13 42 09](https://github.com/facebook/stylex/assets/66249/e760876f-5adb-4d05-85da-3b8e37842e72)

**Reply by nmn** [reply]
@steida If this graphic is correct, what I was referring to was `cap - ex` or the height from the top of `g` to the top of `H`.

The terminology gets a bit mixed depending on what context you are in. I was using typography design terminology and not CSS terminology which is what this graphic seems to be using.

Based on some quick pixel measuring, the "ascender" (top of H - top of g) is about 28px on my screen, while the descender is about 30px on my screen. Pretty close. But then `H` is not the right representative letter as a capital `I` or `l` might be a bit taller.

**Reply by nmn** [reply]
@predaytor You can just look at the CSS output from the tool and translate that to StyleX. We support `::before` and `::after`. You can also just look at the ratio between the `::before` and `::after` margins and use that with math. That's essentially all you need in addition to the blog posts you linked. Instead of a simple `/ 2` as used in those posts, you need to divide in a given ratio based on the font. For `Arial`, it's close `0.5 / 0.5` so `/ 2 ` works fine. For other fonts, you might need different ratios.

**Reply by steida** [reply]
@nmn Thinking out loud, can't it be that radio dynamically computed leveraging alingItem baseline? Browsers know it. Otherwise, they couldn't align different font sizes to one baseline, and if they do that, can we compute it? Or, crazy thinking, what if a whole design would be based on an alingItem baseline? What do you think? 🤔

**Reply by nmn** [reply]
> Browsers know it. Otherwise, they couldn't align different font sizes to one baseline.

This is something I've been nerding out for a while now, and what you're describing is the holy grail I've been chasing for good typography. We're getting very close to being able to do this in the browser with support for the new `lh` unit in CSS. But as far as I can tell, we still need that ratio.

So we have the following relevant units that could be used.
- `lh` is equal to the `line-height` as a length. In CSS `line-height` in the distance from one baseline to the next baseline.
- `ex` is the `x-height` of the first font in the container as a length
- `cap` is the cap height of the first font in the container as a length

With all these units, we should have all the information to calculate what we need:
- We know that the total extra padding is `(1lh - 1cap)`.
- What we don't know is what part of that is used on the top vs bottom.
- My best guess is that this has something to do with the ratio of ascender to descender. i.e `(1cap - 1ex) :: (1em - 1cap)`
  - If we figure out the ratio, we can use [this trick](https://dev.to/janeori/css-type-casting-to-numeric-tanatan2-scalars-582j) using `tan(atan2(a, b))` to divide two lengths in CSS.

Anyway, here's a codepen where I've been experimenting with this:
https://codepen.io/naman34/pen/PoLRGGV

So far the *only* solution that I know works 100% of the time is to wrap every word in a `<span>`. Set `lineHeight: '1cap'` and then use a flex container with flex-wrap to contain all of these spans. You can use `gap` to set a real line-height. This is obviously a horrible solution in most cases, but might be ok for headings.

**Reply by steida** [reply]
@nmn If it's OK for headings, then we could do the whole solution, I guess. I'm working on something like Apple Calendar mixed with Apple Notes, but open source and local-first (with my https://github.com/evoluhq/evolu). Off-topic: I use two baselines, one optimized for content typography and the second optimized for minimal hit region (44), and the baseline is CSS var defined by StyleX Theme, an awesome feature btw.

There are no headings in the app part, only base texts and smaller ones. I push smaller texts to the baseline with EnsureBaseline component rendering invisible 0. No ratio is required because I don't align text to divs. I just want the same baseline.

 ```tsx
/**
 * CSS baseline is defined by font size. This component forces texts with
 * smaller font sizes to have the same baseline.
 */
export const EnsureBaseline = memo(function EnsureBaseline() {
  return (
    <Text aria-hidden style={styles.text}>
      0
    </Text>
  );
});

const styles = stylex.create({
  text: {
    visibility: "hidden",
    marginLeft: "-1ch",
  },
});

<Text aria-hidden key={day.toString()} style={styles.narrowDayName}>
  {day.toLocaleString(undefined, { weekday: "narrow" })}
  <EnsureBaseline />
</Text>

baselineGrid: {
  backgroundImage: `linear-gradient(to bottom, ${colors.hoverAndFocusBackground} 0, transparent 1px)`,
  backgroundRepeat: "repeat-y",
  backgroundSize: `100% ${baselines.half}`,
  // backgroundSize: `100% ${baselines.one}`,
  /**
   * Simulate typographic baseline (a baseline aligned directly under a text).
   * Remember, such a feature is only partially possible without capsize.js
   * because we can use an align-items baseline consistently for only one font
   * size (fontSizes.step0 and smaller). We don't want to use capsize.js
   * because we prefer system fonts. This article is wrong:
   * https://maketypework.com/web-typography-baseline-grids-made-easy. Check
   * my comment there. This ad-hoc offset is for the app baseline only.
   */
  backgroundPositionY: 6.5,
},
 ```
 
`backgroundPositionY` is only for visual testing. 

![Screenshot 2024-02-01 at 9 23 58](https://github.com/facebook/stylex/assets/66249/07cbc40a-a2fd-48c1-8b12-c3149a466a1a)

**Reply by nmn** [reply]
When building a product a very easy solution is to use `cap` for lineHeights. As long as the lineHeight is set with the `cap` unit, `margin-block: calc((1cap - 1lh) / 2);` always works.


Scratch that: `margin-block: calc((1cap - 1lh) / 2);` is good enough in general. Just make a text component with it!
